index.js

```diff
@@ -83,6 +94,116 @@ function attachToBody (options, req, reply, next) {
+  // only a single {} special ending is allowed
+  else if (tokens.filter(t => t === '{}').length > 1) isValid = false
+  // {}, if present, must be the last token
+  else if (tokens.find(t => t === '{}') && tokens[tokens.length - 1] !== '{}') isValid = false
+
+  let hasOpenBracket = false
+  for (let i = 1; i < tokens.length; i++) {
+    const token = tokens[i]
+    const nextToken = tokens[i + 1]
+    // invalid closing bracket
+    if (!hasOpenBracket && token === ']') {
+      isValid = false
+      break
+    }
+    // invalid opening bracket
+    if (hasOpenBracket && token === '[') {
+      isValid = false
+      break
+    }
+    hasOpenBracket = token === '[' || (hasOpenBracket && token !== ']')
+    // orphan opening bracket
+    if (hasOpenBracket && nextToken === undefined) isValid = false
+    // {} is not allowed within brackets
+    if (hasOpenBracket && token === '{}') isValid = false
+    // after a closing bracket, only an opening bracket or {} are allowed
+    if (token === ']' && nextToken && !['[', '{}'].includes(nextToken)) {
+      isValid = false
+      break
+    }
+  }
+
+  // when the {} special ending is supplied, the value must be valid JSON
+  if (isValid && tokens[tokens.length - 1] === '{}') {
+    try {
+      secureJSON.parse(value)
+    } catch (e) {
+      return 'the {} special ending was used, but value is not a valid JSON'
+    }
+  }
+
+  if (tokens.includes('__proto__')) {
+    return '__proto__ is not allowed as a key name'
+  }
+
+  if (tokens.includes('constructor')) {
+    return 'constructor is not allowed as a key name'
+  }
+
+  return isValid ? null : 'invalid serialized key'
+}
+
+function getSerializedKeySegments (options, body, tokens, value) {
+  const segments = [{ parentRef: body, key: tokens[0] }]
+  const parseArrays = options.enableSpecialNotationKeys !== 'objectsOnly'
+
+  for (let i = 1; i < tokens.length; i++) {
+    const lastSegment = segments[segments.length - 1]
+    const token = tokens[i]
+    const prevToken = tokens[i - 1]
+    let parentRef
+    let key
+
+    if (token === ']' && prevToken === '[') {
+      parentRef = lastSegment.parentRef[lastSegment.key] ?? (parseArrays ? [] : {})
+      key = parseArrays ? parentRef.length : Object.keys(parentRef).length
+    } else if (parseArrays && token === ']' && isValidIndex(prevToken)) {
+      parentRef = lastSegment.parentRef[lastSegment.key] ?? (parseArrays ? [] : {})
+      key = +prevToken
+    } else if (token === ']') {
+      parentRef = lastSegment.parentRef[lastSegment.key] ?? {}
+      key = prevToken
+    } else if (token === '{}') {
+      value = secureJSON.parse(value)
+    }
+
+    if (parentRef) segments.push({ parentRef, key })
+  }
+
+  segments[segments.length - 1].value = value
+
+  return segments
+}
+
+function isValidIndex (str) {
+  if (str.length > 1 && str[0] === '0') return false
+  return /^\d+$/.test(str)
+}
+
+function setSerializedKey (segments) {
+  segments.reverse()
```

Comment:
instead of reversing, it would make more sense, to change the for loop

maybe something like this. not tested. possibly wrong.
```js
function setSerializedKey (segments) {
  const len = segments.length

  for (let i = len; i >= 0; --i) {
    const segment = segments[i]
    const prevSegment = segments[i - 1]
    value = prevSegment ? prevSegment.parentRef : value
    segment.parentRef[segment.key] = value
  }
}
```